pkg/cloud/services/networking/router.go

```diff
@@ -190,7 +200,26 @@ func (s *Service) setRouterExternalIPs(openStackCluster *infrav1.OpenStackCluste
 }
 
 func (s *Service) DeleteRouter(openStackCluster *infrav1.OpenStackCluster, clusterName string) error {
-	router, subnet, err := s.getRouter(clusterName)
+	listOpts := routers.ListOpts{}
+	existingRouter := false
+	if openStackCluster.Spec.Router != nil {
+		listOpts = routers.ListOpts(*openStackCluster.Spec.Router)
+		existingRouter = true
+	} else {
+		listOpts.Name = getRouterName(clusterName)
+	}
```

Comment:
Instead of else clause you could have put `Name` directly on `listOpts` initialization.